pop_tools/budget.py

```diff
@@ -0,0 +1,283 @@
+import datetime
+import glob
+
+import numpy as np
+import xarray as xr
+
+from .grid import get_grid
+
+NMOLS_TO_MOLYR = 1e-9 * 86400.0 * 365.0
+
+
+def _process_grid(grid):
+    """Load in POP grid and process coordinates required for budget calculation."""
+    grid_f = get_grid(grid)
+    dz = grid_f.dz.drop('z_t')
+    area = (grid_f.TAREA).rename('area')
+    vol = (area * dz).rename('vol')
+    z = (grid_f.z_w_bot.drop('z_w_bot')).rename('z')
+    mask = grid_f.REGION_MASK.rename('mask')
+    return xr.merge([area, vol, z, mask])
+
+
+def _compute_kmax(budget_depth, z):
+    """Compute the k-index of the maximum budget depth"""
+    kmax = (z <= budget_depth).argmin() - 1
+    kmax = kmax.values
+    # This only occurs when the full depth is requested (or `budget_depth` is greater than the maximum POP depth`)
+    if kmax == -1:
+        kmax = None
+    return kmax
+
+
+def _convert_to_tendency(ds, normalizer, sign=1, kmax=None):
+    """Converts from volume- and area-dependent units to tendencies"""
+    if (kmax is not None) and ('z_t' in normalizer.dims):
+        normalizer = normalizer.isel(z_t=slice(0, kmax + 1))
+    ds = ds * normalizer * sign
+    ds.attrs['units'] = 'nmol/s'
+    return ds
+
+
+def _convert_units(ds):
+    """Converts from nmol/s to mol/yr and adds attribute label"""
+    ds *= NMOLS_TO_MOLYR
+    ds.attrs['units'] = 'mol/yr'
+    return ds
+
+
+def _compute_horizontal_divergence(da, mask, direction=None):
+    """Computes divergence of a tracer flux in the horizontal
+
+    Parameters
+    ----------
+    da : `xarray.DataArray`
+      DataArray with the tracer flux for which the divergence is being computed.
+    direction : str
+      Direction to compute divergence ('zonal' or 'meridional')
+    """
+    if direction == 'zonal':
+        # Moves mask to the left of the region to compute divergence of flow
+        # entering the volume.
+        mask_roll = mask.roll(nlon=-1, roll_coords=True)
+        div = da.where(mask_roll).roll(nlon=1, roll_coords=True) - da.where(mask)
+    elif direction == 'meridional':
+        mask_roll = mask.roll(nlat=-1, roll_coords=True)
+        div = da.where(mask_roll).roll(nlon=1, roll_coords=True) - da.where(mask)
```

Comment:
Note to self -- bug introduced here on refactoring. Should be `.roll(nlat=1, ...)`